schemas/compat-data.schema.json

```diff
@@ -262,7 +262,7 @@
       "errorMessage": {
         "additionalProperties": "Feature names can only contain alphanumerical characters or the following symbols: _-$@"
       },
-      "tsType": "Record<string, Identifier> & {__compat?: CompatStatement};"
+      "tsType": "{[key: string]: Identifier} & {__compat?: CompatStatement};"
```

Comment:
The TypeScript you're suggesting has broken imports of the data on Yari, because it's trying to map `__compat` to an `Identifier` type.  Are you sure this is the right way to do this?

Reply:
This is just a manual definition of `Record<string, Identifier>`. It's semantically no different that `Record<>`, except it allows recursive reference. Note that this does not fail the test added by #16685.

Reply:
With this TypeScript type, however, we run into the following error on Yari:

<img width="1137" alt="Screen Shot 2022-06-16 at 15 34 11" src="https://user-images.githubusercontent.com/5179191/174189459-d5e1f8a7-821c-46b5-ad4b-2a6e78f5f494.png">

Reply:
Tried to build Yari but it's broken in another way 🤔

```
$ npm run build

> @mdn/yari@0.1.6 build
> cross-env NODE_ENV=production node build/cli.js

/mnt/c/Users/Kagami/Documents/GitHub/yari/node_modules/fdir/src/api/walker.js:48
  const needsSeperator = path[path.length - 1] !== sep;
                                   ^
builder/apiBuilder.js:18:10)
    at allDocumentPathsAsTree (/mnt/c/Users/Kagami/Documents/GitHub/yari/content/document-paths.js:20:23)
    at initAllDocumentsPathsTree (/mnt/c/Users/Kagami/Documents/GitHub/yari/content/document-paths.js:85:23)
    at Object.<anonymous> (/mnt/c/Users/Kagami/Documents/GitHub/yari/content/document-paths.js:95:33)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Module._load (node:internal/modules/cjs/loader:827:12)

Node.js v18.1.0
npm notice
npm notice New minor version of npm available! 8.8.0 -> 8.12.2
npm notice Changelog: https://github.com/npm/cli/releases/tag/v8.12.2
npm notice Run npm install -g npm@8.12.2 to update!
npm notice
```

Reply:
Good old code problems left and right... 🙃 

I have a pull request open that uses an updated version of BCD (which includes the error I screenshotted up above), if you'd like to try https://github.com/mdn/yari/pull/6490 out?

Reply:
Just looked at the failure log in https://github.com/mdn/yari/runs/6906221097?check_suite_focus=true, are you sure Yari is blocked by this? That error looks different:

```
 TS2322: Type 'CompatStatement | undefined' is not assignable to type 'CompatStatement'.
  Type 'undefined' is not assignable to type 'CompatStatement'.
    51 |       features.push({
    52 |         name: parentName ? `${parentName}.${subName}` : subName,
  > 53 |         compat: (subIdentifier as BCD.Identifier).__compat,
       |         ^^^^^^
    54 |         depth: depth + 1,
    55 |       });
    56 |       features.push(...listFeatures(subIdentifier, subName, "", depth + 1));
```

That's fixable:

```diff
diff --git a/client/src/document/ingredients/browser-compatibility-table/utils.ts b/client/src/document/ingredients/browser-compatibility-table/utils.ts
index 463c3c867..cd4964915 100644
--- a/client/src/document/ingredients/browser-compatibility-table/utils.ts
+++ b/client/src/document/ingredients/browser-compatibility-table/utils.ts
@@ -50,10 +50,10 @@ export function listFeatures(
     if (subName !== "__compat" && (subIdentifier as BCD.Identifier).__compat) {
       features.push({
         name: parentName ? `${parentName}.${subName}` : subName,
-        compat: (subIdentifier as BCD.Identifier).__compat,
+        compat: (subIdentifier as BCD.Identifier).__compat!,
         depth: depth + 1,
       });
-      features.push(...listFeatures(subIdentifier, subName, "", depth + 1));
+      features.push(...listFeatures(subIdentifier as BCD.Identifier, subName, "", depth + 1));
     }
   }
   return features;
```

This made `yarn start` work.

Reply:
(`__compat!` is probably not a real fix, but anyway)

Reply:
You'll want to check the logs in the "Testing Yari" CI test: https://github.com/mdn/yari/runs/6906221112?check_suite_focus=true

(I'm aware of the other two issues reported by the tests, I plan to fix those at a later time, once this issue is resolved!)

Reply:
Hmm, I got what happened on Yari with your trial to fix:

![image](https://user-images.githubusercontent.com/3396686/174202144-d7ec304e-62f6-4b4a-af09-5961703d258e.png)

It just made it to `any`, and that was hidden by `skipLibCheck` in Yari's tsconfig. 🙂

Reply:
Alright, I'm going to go ahead and merge this PR as-is with the reverted change, and then going to follow-up with updating the default TS export to cast to `any` first.  I understand that's very bad to do and may result in unintended typecasts, but at this point, TypeScript users cannot use newer versions of BCD, including Yari.